ide/rust/SyncRunner.kt

```diff
@@ -0,0 +1,128 @@
+import com.vaticle.bazel.distribution.common.Logging.Logger
+import com.vaticle.bazel.distribution.common.shell.Shell
+import com.vaticle.bazel.distribution.common.util.FileUtil.listFilesRecursively
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.ASPECTS
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.BAZEL
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.BAZEL_BIN
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.BUILD
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.INFO
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.OUTPUT_BASE
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.OUTPUT_GROUPS_RUST_IDE_SYNC
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.QUERY
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.RUST_IDE_SYNC_ASPECT
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.RUST_TARGETS_DEPS_QUERY
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.RUST_TARGETS_QUERY
+import com.vaticle.dependencies.ide.rust.SyncRunner.ShellArgs.VATICLE_REPO_PREFIX
+import picocli.CommandLine
+import java.io.File
+import java.nio.file.Path
+import java.util.concurrent.Callable
+import kotlin.io.path.Path
+import kotlin.system.exitProcess
+
+fun main(args: Array<String>): Unit = exitProcess(CommandLine(SyncRunner()).execute(*args))
+
+@CommandLine.Command(name = "sync", mixinStandardHelpOptions = true)
+class SyncRunner : Callable<Unit> {
+
+    @CommandLine.Option(names = ["--verbose", "-v"], required = false)
+    var verbose: Boolean = false
+
+    lateinit var logger: Logger
+    private lateinit var shell: Shell
+    private val buildWorkspaceDir = Path(System.getenv("BUILD_WORKSPACE_DIRECTORY"))
+    lateinit var bazelOutputBase: Path
+
+    override fun call() {
+        logger = Logger(logLevel = if (verbose) DEBUG else ERROR)
+        shell = Shell(logger, verbose)
+        bazelOutputBase = Path(shell.execute(listOf(BAZEL, INFO, OUTPUT_BASE), buildWorkspaceDir).outputString().trim())
+        val rustTargets = rustTargets(buildWorkspaceDir)
+        validateBuildWorkspace(rustTargets)
+        loadRustToolchainAndExternalDeps(rustTargets)
+        rustRepoPaths().forEach { sync(it) }
+    }
+
+    private fun validateBuildWorkspace(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets + "--build=false", baseDir = buildWorkspaceDir)
+    }
+
+    private fun loadRustToolchainAndExternalDeps(rustTargets: List<String>) {
+        shell.execute(command = listOf(BAZEL, BUILD) + rustTargets, baseDir = buildWorkspaceDir, throwOnError = false)
+    }
+
+    private fun rustTargets(repoPath: Path): List<String> {
+        return shell.execute(listOf(BAZEL, QUERY, RUST_TARGETS_QUERY), repoPath)
+            .outputString().split("\n").filter { it.isNotBlank() }
+    }
+
+    private fun rustRepoPaths(): List<Path> {
```

Comment:
This method does feel quite terse, therefore let's make it less so. Also the name can be more descriptive. Something like `vaticleRustRepo()` perhaps?

Reply:
Renamed to `vaticleRustRepoPaths`. What do you mean by "let's make it less so"?